test/unit/master/test_build.py

```diff
@@ -31,27 +33,54 @@ def test_allocate_slave_calls_slave_setup(self):
 
-    def test_slave_doesnt_use_more_than_max_executors(self):
+    def test_build_doesnt_use_more_than_max_executors(self):
         subjobs = self._create_subjobs()
         mock_project_type = self._create_mock_project_type()
         fake_setup_command = 'mock command'
         mock_slaves = [self._create_mock_slave(num_executors=5) for _ in range(3)]
-        expected_num_max_executors = 12  # We expect the slave to use 12 out of 15 available executors.
+        expected_num_executors = 12  # We expect the build to use 12 out of 15 available executors.
 
         build = Build(BuildRequest({'setup': fake_setup_command}))
         build.execute_next_subjob_on_slave = MagicMock()
 
-        build.prepare(subjobs, mock_project_type, self._create_job_config(max_executors=expected_num_max_executors))
+        build.prepare(subjobs, mock_project_type, self._create_job_config(max_executors=expected_num_executors))
         [build.allocate_slave(mock_slave) for mock_slave in mock_slaves]
         [build.begin_subjob_executions_on_slave(mock_slave) for mock_slave in mock_slaves]
 
-        self.assertEqual(build.execute_next_subjob_on_slave.call_count, expected_num_max_executors,
+        self.assertEqual(build.execute_next_subjob_on_slave.call_count, expected_num_executors,
+                         'Build should start executing as many subjobs as its max_executors setting.')
+
+    def test_build_doesnt_use_more_than_max_executors_per_slave(self):
+        subjobs = self._create_subjobs()
+        mock_project_type = self._create_mock_project_type()
+        fake_setup_command = 'mock command'
+        mock_slaves = [self._create_mock_slave(num_executors=5) for _ in range(3)]
+        max_executors_per_slave = 2
+        expected_total_num_executors_used = 6  # We expect the build to use 2 executors on each of the 3 slaves.
+
+        build = Build(BuildRequest({'setup': fake_setup_command}))
+        build.execute_next_subjob_on_slave = MagicMock()
+
+        build.prepare(subjobs, mock_project_type,
+                      self._create_job_config(max_executors_per_slave=max_executors_per_slave))
+        [build.allocate_slave(mock_slave) for mock_slave in mock_slaves]
+
+        expected_current_num_executors_used = 0
+        for i in range(len(mock_slaves)):
+            build.begin_subjob_executions_on_slave(mock_slaves[i])
+            expected_current_num_executors_used += max_executors_per_slave
+            self.assertEqual(
+                build.execute_next_subjob_on_slave.call_count, expected_current_num_executors_used,
+                'After allocating {} slaves, build with max_executors_per_slave set to {} should only be using {} '
+                'executors.'.format(i + 1, max_executors_per_slave, expected_current_num_executors_used))
+
+        self.assertEqual(build.execute_next_subjob_on_slave.call_count, expected_total_num_executors_used,
```

Comment:
This last assertion is redundant, but I think it makes the test clearer. 

Just realized I need to fix the message for this assertion though. Will do.